### PR TITLE
Meshlet software raster + start of cleanup

### PR DESCRIPTION
**Pull request by JMS55**
# Objective
- Faster meshlet rasterization path for small triangles
- Avoid having to allocate and write out a triangle buffer
- Refactor gpu_scene.rs

## Solution
- Replace the 32bit visbuffer texture with a 64bit visbuffer buffer, where the left 32 bits encode depth, and the right 32 bits encode the existing cluster + triangle IDs. Can't use 64bit textures, wgpu/naga doesn't support atomic ops on textures yet.
- Instead of writing out a buffer of packed cluster + triangle IDs (per triangle) to raster, the culling pass now writes out a buffer of just cluster IDs (per cluster, so less memory allocated, cheaper to write out). 
  - Clusters for software raster are allocated from the left side
  - Clusters for hardware raster are allocated in the same buffer, from the right side
  - The buffer size is fixed at MeshletPlugin build time, and should be set to a reasonable value for your scene (no warning on overflow, and no good way to determine what value you need outside of renderdoc - I plan to fix this in a future PR adding a meshlet stats overlay)
  - Currently I don't have a heuristic for software vs hardware raster selection for each cluster. The existing code is just a placeholder. I need to profile on a release scene and come up with a heuristic, probably in a future PR.
  - The culling shader is getting pretty hard to follow at this point, but I don't want to spend time improving it as the entire shader/pass is getting rewritten/replaced in the near future.
- Software raster is a compute workgroup per-cluster. Each workgroup loads and transforms the <=64 vertices of the cluster, and then rasterizes the <=64 triangles of the cluster.
  - Two variants are implemented: Scanline for clusters with any larger triangles (still smaller than hardware is good at), and brute-force for very very tiny triangles
  - Once the shader determines that a pixel should be filled in, it does an atomicMax() on the visbuffer to store the results, copying how Nanite works
  - On devices with a low max workgroups per dispatch limit, an extra compute pass is inserted before software raster to convert from a 1d to 2d dispatch (I don't think 3d would ever be necessary).
  - I haven't implemented the top-left rule or subpixel precision yet, I'm leaving that for a future PR since I get usable results without it for now
  - Resources used: https://kristoffer-dyrkorn.github.io/triangle-rasterizer and chapters 6-8 of https://fgiesen.wordpress.com/2013/02/17/optimizing-sw-occlusion-culling-index
- Hardware raster now spawns 64*3 vertex invocations per meshlet, instead of the actual meshlet vertex count. Extra invocations just early-exit. 
  - While this is slower than the existing system, hardware draws should be rare now that software raster is usable, and it saves a ton of memory using the unified cluster ID buffer. This would be fixed if wgpu had support for mesh shaders.
  - Instead of writing to a color+depth attachment, the hardware raster pass also does the same atomic visbuffer writes that software raster uses. 
  - We have to bind a dummy render target anyways, as wgpu doesn't currently support render passes without any attachments
  - Material IDs are no longer written out during the main rasterization passes.
  - If we had async compute queues, we could overlap the software and hardware raster passes.
- New material and depth resolve passes run at the end of the visbuffer node, and write out view depth and material ID depth textures

### Misc changes
- Fixed cluster culling importing, but never actually using the previous view uniforms when doing occlusion culling
- Fixed incorrectly adding the LOD error twice when building the meshlet mesh
- Splitup gpu_scene module into meshlet_mesh_manager, instance_manager, and resource_manager
  - resource_manager is still too complex and inefficient (extract and prepare are way too expensive). I plan on improving this in a future PR, but for now ResourceManager is mostly a 1:1 port of the leftover MeshletGpuScene bits.
- Material draw passes have been renamed to the more accurate material shade pass, as well as some other misc renaming (in the future, these will be compute shaders even, and not actual draw calls)

---

## Migration Guide
- TBD (ask me at the end of the release for meshlet changes as a whole)

### CODE REVIEW

**Review by atlv24 — COMMENTED** [review]
This is insanely impressive.

I haven't reviewed the raster code in depth yet, but I really like the split of gpu_scene. I cant wait for render graph rework to hopefully reduce the amount of boilerplate.

I'm gonna mess around with it locally for a while, in particular the raster and culling, before hitting approve.

File: crates/bevy_pbr/src/meshlet/downsample_depth.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/downsample_depth.wgsl]: Done

File: crates/bevy_pbr/src/meshlet/resource_manager.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/resource_manager.rs:49]: Need to remove `previous_depth_pyramids` and just cache a struct per-view holding every single per-view resource and bind group, and only recreate it when the view changes. Just a pain to implement, haven't felt like doing it yet (hence why resource_manager.rs is so gross). I don't want to block the PR on it.

File: crates/bevy_pbr/src/meshlet/resource_manager.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/resource_manager.rs:377]: Yeah it's correct. The whole file is confusing and slow, needs rewriting, see my other comment and the PR description.

File: crates/bevy_pbr/src/meshlet/visibility_buffer_raster_node.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/visibility_buffer_raster_node.rs:106]: This is existing code, but each workgroup processes 128 clusters, so we need `x = ceil(N / 128)` workgroups total. Of course, since we want one dispatch over every single cluster (millions across all instances/LODs) in the scene, we'd quickly hit the ~65,536 per-dispatch-dimension limit if we tried to dispatch 3 million workgroups in only the x dimension. Can't do a 1d dispatch.

Using the ceil of the cuberoot lets us spread the dispatch across all 3 x/y/z dimensions to get around the limit. In the shader, global_workgroup_id is converted back to a 1d index. 

This is the same trick as the remap_1d_to_2d_dispatch shader, which does the same thing on the GPU for converting a 1d indirect dispatch to a 2d dispatch (3d is unlikely to ever be necessary).

This is going to be irrelevant once I switch to persistent threads-style culling of clusters, as then we just need (1?) a low, fixed number of workgroups to process the scene no matter the cluster count.

The rename from scene_meshlet_count to scene_cluster_count is just a rename that was missed when making the terminology more consistent in past PRs, it's not functionally different.

File: crates/bevy_pbr/src/meshlet/meshlet_mesh_manager.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/meshlet_mesh_manager.rs]: Done.

**Comment by JMS55** [comment]
> I'm gonna mess around with it locally for a while, in particular the raster and culling, before hitting approve.

Please do. Occlusion culling is probably still bugged, may want to disable that. But definitely check that the culling and software raster code is correct (sorry the culling code is so gross, it'll be rewritten entirely once I switch to persistent threads-style culling). Everything else in this PR is just CPU boilerplate or adapting simple GPU passes to the new visbuffer.

Something that needs testing if you have time/motivation is "Test if mixing forward+prepass and deferred rendering when mixing MeshletMesh and Mesh entities is still broken". The new depth resolve pass after the visbuffer raster should hopefully fix this, so that I can remove the caveat from the module docs.

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:87]: do you also need to `min_x = max(0, min_x)` etc?

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:111]: what about max_y - min_y being huge? maybe its better to use triangle_double_area?

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:115]: why `1e8`?

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:123]: i'd move this `+ min_x` down to be next to the other one just so its clearer that both x0 and x1 are both shifted

File: crates/bevy_pbr/src/meshlet/cull_clusters.wgsl
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/cull_clusters.wgsl:95]: why `var`?

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:131]: that's annoying, i'll see if i can figure this out but dont block merge on this

**Review by atlv24 — APPROVED** [review]

File: crates/bevy_pbr/src/meshlet/cull_clusters.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/cull_clusters.wgsl:95]: Reassigned later on lines 133+134

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl]: Done

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:115]: Idk, it's what the nanite slides do.

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:111]: Not really worth doing the whole setup for the scanline calcs in both dimensions. Only worth doing for the inner loop

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:87]: Iirc the u32(foo) should turn all negatives into 0.

File: crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl:115]: I tried to understand what this is doing, but never quite figured it out.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (IceSentry) [inline crates/bevy_pbr/src/meshlet/from_mesh.rs:292]: Is this about the bevy mesh api or more specific to meshlets?

File: crates/bevy_pbr/src/meshlet/instance_manager.rs
Comment (IceSentry) [inline crates/bevy_pbr/src/meshlet/instance_manager.rs:20]: Maybe it's fine with the name space, but that feels a bit generic? Why not something like `VirtualGeometryMeshManager`? To be clear, this isn't a blocker or anything, I'm just curious.

File: crates/bevy_pbr/src/meshlet/from_mesh.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/from_mesh.rs:292]: Specific to meshlets. Simplification (i.e. the LOD building) only accounts for position atm, and not things like UVs/normals. See https://github.com/zeux/meshoptimizer/issues/158.

File: crates/bevy_pbr/src/meshlet/instance_manager.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/instance_manager.rs:20]: Eh I didn't want to get hung up on naming. We can change it later.

File: crates/bevy_pbr/src/meshlet/meshlet_mesh_manager.rs
Comment (IceSentry) [inline crates/bevy_pbr/src/meshlet/meshlet_mesh_manager.rs:20]: Maybe this should have Gpu somewhere in the name to clarify it's puprose?

File: crates/bevy_pbr/src/meshlet/resource_manager.rs
Comment (IceSentry) [inline crates/bevy_pbr/src/meshlet/resource_manager.rs:136]: That's a lot of bindings, but that does feel a lot easier to read then when we were still using the raw structs 😅

**Review by IceSentry — APPROVED** [review]
Okay, I'm not seeing anything obviously wrong. I confirmed that the code doesn't break anything else. The various `*Manager` structs are a bit generically named, but that doesn't really matter for now.

I get a clippy lint locally but it's in a crate that isn't modified in this PR so not sure what's up with that. Other than that LGTM.

File: crates/bevy_pbr/src/meshlet/resource_manager.rs
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/resource_manager.rs:136]: Yeahhh I want automated bind groups in the render graph for a reason... Arguably this explicitness is better for perf, but it can't be that much slower to hash-and-cache, and this sucks to write and tweak.

**Comment by mockersf** [comment]
this PR maybe made things faster, but it also greatly reduces the range of hardware where it works. meshlets used to work on macOS or on Vulkan with software renderer, not anymore.

**Comment by JMS55** [comment]
I'm not open to supporting GPUs without 64bit atomics (meaning no software raster).

Besides being an insane amount of extra code to support (pretty much an entire extra copy of the codebase), there's no point in telling artists to design their scenes around this feature for pixel level geometry detail, and then have older users try to use it, completely fall over due to lack of software raster + low vram, and then blame the game developers for not optimizing their game when the developers can't do anything besides making an entire second copy of the scene designed for working without meshlets. This is not a feature meant for older platforms, and the only reason it ran before was a temporary setup so that I could develop the rest of the feature without waiting for wgpu to implement 64bit atomics.

It should run on DX12 (SM 6.6+ iirc) and Metal (M2+ iirc), except wgpu/naga is bugged on DX12/Metal with these shaders.

**Comment by JMS55** [comment]
For my future reference, another good SW raster tutorial: https://web.archive.org/web/20050408192410/http://sw-shader.sourceforge.net/rasterizer.html

**Pull request merged**
